app/multitenant/memcache_client.go

```diff
@@ -0,0 +1,131 @@
+package multitenant
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"sort"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/bradfitz/gomemcache/memcache"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/weaveworks/scope/report"
+)
+
+var (
+	memcacheCounter = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "scope",
+		Name:      "memcache",
+		Help:      "Reports that missed our in-memory cache but went to our memcache",
+	}, []string{"result"})
+
+	memcacheRequestDuration = prometheus.NewSummaryVec(prometheus.SummaryOpts{
+		Namespace: "scope",
+		Name:      "memcache_request_duration_nanoseconds",
+		Help:      "Time spent doing memcache requests.",
+	}, []string{"method", "status_code"})
+)
+
+func init() {
+	prometheus.MustRegister(memcacheCounter)
+	prometheus.MustRegister(memcacheRequestDuration)
+}
+
+type memcacheClient struct {
+	client     *memcache.Client
+	expiration int32
+}
+
+// Create a new MemcacheClient that gets its server list from SRV and updates
+// the server list on a regular basis.
+func NewMemcacheClient(host string, timeout time.Duration, service string, updateInterval time.Duration, expiration int32) (*memcacheClient, error) {
+	var servers memcache.ServerList
```

Comment:
Updated for your comfort (although I'd like to know more about why you feel that way).

Is calling `NewMemcacheClient` safe to call from goroutines? It doesn't mutate any non-local state, so I don't see why not. Or do you mean, is the goroutine in `NewMemcacheClient` safe? Yes, it only mutates `ServerList`, which is documented as being safe to call from goroutines.